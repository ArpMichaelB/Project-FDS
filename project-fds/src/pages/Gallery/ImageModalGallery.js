import React, { Component } from "react";
import ImageGallery from "react-image-gallery";
import PropTypes from "prop-types";

class ImageModalGallery extends Component {

	static propTypes = {
		gallery_images:PropTypes.array,
		index:PropTypes.number
	}

	componentDidMount(){
		this.imageGallery.slideToIndex(this.props.index);
	}

	componentDidUpdate(){
		this.imageGallery.slideToIndex(this.props.index);
	}

	render(){
		return(
			<ImageGallery ref={i => this.imageGallery = i} items={this.props.gallery_images} lazyLoad={true} showBullets={false} showThumbnails={false} showPlayButton={false} />
		);
	}

}

export default ImageModalGallery;