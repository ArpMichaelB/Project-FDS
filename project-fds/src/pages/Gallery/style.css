.cards{
    width: 300px;
    display: inline-block;
}

.icon{
    font-size: 25px;
    position: absolute;
    top: 50%;
}

.small_image{
    width: 300px;
    height: 300px;
}

#left-icon{
    left:0;
}

#right-icon{
    right:0;
}