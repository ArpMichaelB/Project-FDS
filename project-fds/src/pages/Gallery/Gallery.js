import React, { Component } from "react";
import { Card, Modal  } from "antd";
import InfiniteScroll from "react-infinite-scroll-component";
import LazyLoad from "react-lazyload";
import ImageModalGallery from "./ImageModalGallery";
import DocumentTitle from "react-document-title";

import "react-image-gallery/styles/css/image-gallery.css";
import "./style.css";

class Gallery extends Component {
	constructor(props){
		super(props);

		this.initialLoad = true;
		this.imagesLoading = true;
		this.lastLocation = "";
		this.images = [];

		this.state={visible:false, width:"", index:0, scoll_index:10, hasMore:true, h3_display:"inline", cards:[]};

		this.cancel=this.cancel.bind(this);
		this.showModal=this.showModal.bind(this);

		fetch("https://shr4ny5edi.execute-api.us-east-1.amazonaws.com/default/image?operation=readAll")
			.then(response => response.json())
			.then(data => {
				this.images = JSON.parse(data.body);

				let scoll_index = this.state.scoll_index + 10;
				if(scoll_index >= this.images.length){
					scoll_index = this.images.length;
				}

				let start = scoll_index-10;
				if(start < 0){
					start = 0;
				}
		
				let cards = [];
				for(let i = start;i < scoll_index;i++){
					if(this.images[i].sitelocation.toLowerCase() !== this.lastLocation){
						this.lastLocation = this.images[i].sitelocation.toLowerCase();
						cards.push(<div className="title"><h2>{this.lastLocation.charAt(0).toUpperCase() + this.lastLocation.slice(1)}</h2></div>);
					}
					cards.push(
						<Card
							hoverable
							key={i}
							bodyStyle={{padding: "0"}}
							className="cards"
							onClick={() => this.showModal(i)}
							cover={
								<LazyLoad placeholder={<img alt="placeholder" className="small_image" src="/gallery/loading_image.gif" />}>
									<img alt={this.images[i].alt} className="small_image" src={"https://s3.amazonaws.com/fdsimagebucket/" + this.images[i].link} />
								</LazyLoad>
							}
						/>
					);
				}

				this.gallery_images = [];

				this.images.forEach(image => {
					this.gallery_images.push({original:"https://s3.amazonaws.com/fdsimagebucket/" + image.link, 
						originalClass:"gallery_image", description:image.caption});
				});

				this.imagesLoading = false;
				this.setState({cards:cards});
			});
	}

	render(){
		return (
			<main className="Gallery">
				<DocumentTitle title="Gallery" />
				<div className="title"><h1>Gallery</h1></div>
				<InfiniteScroll
					dataLength={this.state.cards.length}
					next={this.loadFunc}
					hasMore={this.state.hasMore}
					loader={<h4>Loading ...</h4>}
				>
					{this.state.cards}
				</InfiniteScroll>
				<Modal visible={this.state.visible} id="modal" onCancel={this.cancel} bodyStyle={{padding:"0"}} 
					footer={null}>
					<ImageModalGallery index={this.state.index} gallery_images={this.gallery_images} />
				</Modal>
				<h3 style={{display:this.state.h3_display}} ref={i => this.bottom_text = i}>End of Gallery</h3>
			</main>
		);
	}

	componentDidUpdate(){
		if(this.initialLoad && !this.imagesLoading){
			let rect = this.bottom_text.getBoundingClientRect();
			let elemTop = rect.top;
			let elemBottom = rect.bottom;
		
			let isVisible = (elemTop >= 0) && (elemBottom <= window.innerHeight);

			if(isVisible){
				if(this.state.scoll_index >= this.images.length){
					this.initialLoad = false;
					this.setState({h3_display:"none"});
				}
				else{
					this.loadFunc();
				}
			}
			else{
				this.initialLoad = false;
				this.setState({h3_display:"none"});
			}
		}
	}

	loadFunc = () => {

		if(this.state.scoll_index >= this.images.length){
			this.setState({hasMore:false});
			return;
		}
		
		let scoll_index = this.state.scoll_index + 10;
		if(scoll_index >= this.images.length){
			scoll_index = this.images.length;
		}

		let start = scoll_index-10;
		if(start < 0){
			start = 0;
		}

		let cards = [];
		for(let i = start;i < scoll_index;i++){
			if(this.images[i].sitelocation.toLowerCase() !== this.lastLocation){
				this.lastLocation = this.images[i].sitelocation.toLowerCase();
				cards.push(<div className="title"><h2>{this.lastLocation.charAt(0).toUpperCase() + this.lastLocation.slice(1)}</h2></div>);
			}
			cards.push(
				<Card
					hoverable
					key={i}
					bodyStyle={{padding: "0"}}
					className="cards"
					onClick={() => this.showModal(i)}
					cover={
						<LazyLoad placeholder={<img alt="placeholder" className="small_image" src="/gallery/loading_image.gif" />}>
							<img alt={this.images[i].alt} className="small_image" src={"https://s3.amazonaws.com/fdsimagebucket/" + this.images[i].link} />
						</LazyLoad>
					}
				/>
			);
		}
		this.setState((prevState) => {return {scoll_index:scoll_index, cards:prevState.cards.concat(cards)};});
	}

	cancel(){
		this.setState({visible:false});
	}

	showModal(index){
		this.setState({
			visible:true, index:index
		});
	}
}

export default Gallery;