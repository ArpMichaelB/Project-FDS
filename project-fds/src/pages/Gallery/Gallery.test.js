/* eslint-disable quotes */
/* eslint-disable no-undef */
import Gallery from "./Gallery";
import { Promise } from "q";
import React from "react";
import Adapter from "enzyme-adapter-react-16";
import { shallow, mount, configure } from "enzyme";
import toJson from "enzyme-to-json";

configure({adapter: new Adapter()});

let dataString = "[";
for(let i = 0;i < 15;i++){
	dataString += '{"link":"","caption":"Bucktown","alt":"Bucktown","sitelocation":"Bucktown","id":' + i + '},';
}
dataString += '{"link":"","caption":"Bucktown","alt":"Bucktown","sitelocation":"Bucktown","id":' + 15 + '}]';

var data ={body:dataString};

const flushPromises = () => {
	return new Promise(resolve => setImmediate(resolve));
};

global.fetch = jest.fn().mockImplementation(() => {
	return Promise.resolve({
		json: () => {
			return Promise.resolve(data);
		}
	});
});

test("constructor should have called fetch", () => {
	shallow(<Gallery />);

	expect(global.fetch).toHaveBeenCalledTimes(1);
	expect(global.fetch).toHaveBeenCalledWith("https://shr4ny5edi.execute-api.us-east-1.amazonaws.com/default/image?operation=readAll");
});

test("cards should be created with 12 elements", async () => {

	let gallery = shallow(<Gallery />);
	await flushPromises();
	gallery.update();

	expect(gallery.state().cards.length).toBe(11);
});

test("clicking card should show modal", async () => {
	let gallery = mount(<Gallery />);
	await flushPromises();
	gallery.update();

	expect(gallery.state().visible).toBe(false);

	let card0 = gallery.find(".ant-card").first();
	card0.props().onClick();

	expect(gallery.state().visible).toBe(true);

	gallery.unmount();
});

test("snapshot test", async () => {
	const gallery = mount(<Gallery />);
	await flushPromises();
	gallery.update();

	let tree = toJson(gallery, {mode: "deep"});
	expect(tree).toMatchSnapshot();

	let card0 = gallery.find(".ant-card").first();
	card0.props().onClick();

	tree = toJson(gallery, {mode: "deep"});
	expect(tree).toMatchSnapshot();

	gallery.unmount();
});