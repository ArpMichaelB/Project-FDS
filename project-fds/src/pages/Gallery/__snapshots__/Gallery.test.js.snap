// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<main
  className="Gallery"
>
  <div
    className="title"
  >
    <h1>
      Gallery
    </h1>
  </div>
  <div
    style={Object {}}
  >
    <div
      className="infinite-scroll-component "
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "height": "auto",
          "overflow": "auto",
        }
      }
    >
      <div
        className="title"
        key="bucktown"
      >
        <h2>
          Bucktown
        </h2>
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
    </div>
  </div>
  <h3>
    End Of Gallery
  </h3>
</main>
`;

exports[`snapshot test 2`] = `
<main
  className="Gallery"
>
  <div
    className="title"
  >
    <h1>
      Gallery
    </h1>
  </div>
  <div
    style={Object {}}
  >
    <div
      className="infinite-scroll-component "
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "height": "auto",
          "overflow": "auto",
        }
      }
    >
      <div
        className="title"
        key="bucktown"
      >
        <h2>
          Bucktown
        </h2>
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
      <div
        className="ant-card cards ant-card-bordered ant-card-hoverable"
        onClick={[Function]}
      >
        <div
          className="ant-card-cover"
        >
          <img
            alt="placeholder"
            className="small_image"
            src="/gallery/loading_image.gif"
          />
        </div>
        <div
          className="ant-card-body"
          style={
            Object {
              "padding": "0",
            }
          }
        />
      </div>
    </div>
  </div>
  <h3>
    End Of Gallery
  </h3>
</main>
`;
